test/new_relic/telemetry_sdk/span_test.rb

```diff
@@ -0,0 +1,115 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-telemetry-sdk-ruby/blob/main/LICENSE for complete details.
+
+require File.expand_path(File.join(File.dirname(__FILE__),'../..','test_helper'))
+
+require 'new_relic/telemetry_sdk/span'
+require 'new_relic/telemetry_sdk/util'
+
+module NewRelic
+  module TelemetrySdk
+    class SpanTest < Minitest::Test
+      def test_required_attributes
+        span = Span.new
+        assert span.id.is_a? String
+        assert span.trace_id.is_a? String
+        assert span.start_time_ms.is_a? Integer
+      end
+
+      def test_recommended_attributes
+        span = Span.new duration_ms: 123456,
+                        parent_id: "c617c2813a222a34",
+                        name: "Name",
+                        service_name: "My Service"
+
+        assert_equal 123456, span.duration_ms
+        assert span.duration_ms.is_a? Integer
+
+        assert_equal "Name", span.name
+        assert span.name.is_a? String
+
+        assert_equal "c617c2813a222a34", span.parent_id
+        assert span.parent_id.is_a? String
+
+        assert_equal "My Service", span.service_name
+        assert span.service_name.is_a? String
+      end
+
+      def test_custom_attributes
+        custom_attributes = {
+          :'user.email' => "me@newr.com",
+          :custom_key   => "custom_value"
+        }
+
+        span = Span.new custom_attributes: custom_attributes
+
+        assert_equal custom_attributes, span.custom_attributes
+      end
+
+      def test_finish_with_end_time_supplied
+        span = Util.stub :time_to_ms, 0 do
+          start_time_ms = Util.time_to_ms
+          Span.new start_time_ms: start_time_ms
+        end
+
+        Util.stub :time_to_ms, 1000 do
+          end_time_ms = Util.time_to_ms
+          span.finish end_time_ms: end_time_ms
+        end
+
+        assert_equal 1000, span.duration_ms
+      end
+
+      def test_finish_without_end_time_supplied
```

Comment:
What about a test where `end_time` supplied is before `start_time`?

Reply:
Good question. If `start_time` and `end_time` result in a negative duration, there is no problem with storing, marshaling, or sending the span, so according to the [Telemetry SDK specs](https://github.com/newrelic/newrelic-telemetry-sdk-specs/blob/master/validation.md#sdk-validation) we shouldn't try to do anything special in that situation.